{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Practical 1 : integrating single-cell and spatial analyses in python\n",
    "In this practical, you will consecutively analyse single-cell (10X genomics chromium) and spatial transcriptomics (10X genomics visium) from a pleural mesothelioma tumor.\n",
    "\n",
    "We will focus on a single donor, patient 3B, a 80+ y.o. male without asbestos exposure on which 1 single-cell and 1 spatial sequencing were performed.\n",
    "\n",
    "The analyses will be performed using scanpy. See https://scanpy.readthedocs.io/en/stable/tutorials/index.html for documentation and tutorials as well as the excellent single-cell best practices workflow (https://www.sc-best-practices.org/cellular_structure/clustering.html) for help on the relevant functions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part I: single-cell analyses\n",
    "\n",
    "The data was already processed (see scripts in files/scripts/practical1/2025/), with QC and filtering of cells and features as well as ambiant RNA correction and doublet flagging.\n",
    "\n",
    "In the fist part of this practical, you will vizualise, clusterise and annotate the dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading modules\n",
    "\n",
    "We will mostly use module scanpy, attributing it the alias sc. scanpy has submodules for preprocessing (submodule pp, accessed using sc.pp), for its main tools functions (submodule tl, accessed with sc.tl), and for plotting (submodule pl, accessed with sc.pl). See the scanpy doc for a full list of functions from each submodule https://scanpy.readthedocs.io/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mangel/miniconda3/envs/meso2_conda_env/lib/python3.12/site-packages/anndata/utils.py:434: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/mangel/miniconda3/envs/meso2_conda_env/lib/python3.12/site-packages/anndata/utils.py:434: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/mangel/miniconda3/envs/meso2_conda_env/lib/python3.12/site-packages/anndata/utils.py:434: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/mangel/miniconda3/envs/meso2_conda_env/lib/python3.12/site-packages/anndata/utils.py:434: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/mangel/miniconda3/envs/meso2_conda_env/lib/python3.12/site-packages/anndata/utils.py:434: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/mangel/miniconda3/envs/meso2_conda_env/lib/python3.12/site-packages/anndata/utils.py:434: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/mangel/miniconda3/envs/meso2_conda_env/lib/python3.12/site-packages/anndata/utils.py:434: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading the data\n",
    "The processed data is in the /data/Training-MG/files/data/Practical1/scRNAseq/2025/processed folder.\n",
    "\n",
    "Q1: load the data using the read function (scanpy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q2: Print the anndata object. How many cells are there? How many features?\n",
    "\n",
    "Q3: What are the different layers present? What do they correspond to?\n",
    "\n",
    "Q4: Plot the distribution of gene expression for gene BAP1 (see https://scanpy-tutorials.readthedocs.io/en/latest/plotting/core.html for plotting options from submodule sc.pl), for the different layers (layer option). Which one would favor for downsteam analysis and why?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clustering and vizualisation\n",
    "\n",
    "Q5: compute the K-nearest neighbour graph (function neighbors from the sc.pp submodule) and leiden clustering function (function leiden from the sc.tl submodule) .\n",
    "\n",
    "Q6: compute (function umap from the scanpy pp submodule) and plot the results using UMAP (function umap from the pl submodule). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Identifying tumor cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q7: Perform copy number calling using module infercnvpy. \n",
    "\n",
    "Q8: Visualise the CNV profile of the tumor. Do you find some of the known alterations driving PM (see e.g. https://www.nature.com/articles/s41588-023-01321-1#Fig3 panel b for genes frequently amplified or deleted, and https://www.genecards.org/ to find their locations) ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotating cells\n",
    "\n",
    "For the first part of the cell annotation, we will use the celltypist module. See https://www.celltypist.org/ for the documentation and tutorials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import celltypist\n",
    "from celltypist import models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Q9: Download the immune single-cell reference \"Immune_All_High.pkl\" using the download_models function. Print the object. What cell types are in the reference? \n",
    "\n",
    "Q10: Perform an annotation using the CellTypist trained classifier for immune cells (you can use the \"coarser\" model)\n",
    "\n",
    "Q11: Check the quality of the annotation. Are there cell types with dubious annotations, and if so, why do you think they are hard to annotate?\n",
    "\n",
    "Q12: plot the proportion of cell types. Does it match your expectations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  },
  "metadata": {
   "interpreter": {
    "hash": "a9be70e7d481e5545e76aa1dcb3fa636e7b090925780b153c2ebb9fca3f8820f"
   }
  },
  "orig_nbformat": 2
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
